lib/storage/index.ts

```diff
@@ -1,13 +1,52 @@
+import * as Logger from '../Logger';
+
 import PlatformStorage from './platforms';
 import InstanceSync from './InstanceSync';
+import NoopProvider from './providers/NoopProvider';
 import type StorageProvider from './providers/types';
 
-const provider = PlatformStorage;
+let provider = PlatformStorage;
 let shouldKeepInstancesSync = false;
+let finishInitalization: (value?: unknown) => void;
+const initPromise = new Promise((resolve) => {
+    finishInitalization = resolve;
+});
 
 type Storage = {
     getStorageProvider: () => StorageProvider;
-} & StorageProvider;
+} & Omit<StorageProvider, 'name'>;
+
+/**
+ * Degrade performance by removing the storage provider and only using cache
+ */
+function degradePerformance(error: Error) {
+    Logger.logAlert(`Error while using ${provider.name}. Falling back to only using cache and dropping storage.`);
+    console.error(error);
+    provider = NoopProvider;
+}
+
+/**
+ * Runs a piece of code and degrades performance if certain errors are thrown
+ */
+function tryOrDegradePerformance<T>(fn: () => Promise<T> | T): Promise<T> {
+    return new Promise<T>((resolve, reject) => {
+        initPromise.then(() => {
+            try {
+                resolve(fn());
+            } catch (error) {
+                // Test for known critical errors that the storage provider throws, e.g. when storage is full
+                if (error instanceof Error) {
+                    // IndexedDB error when storage is full (https://github.com/Expensify/App/issues/29403)
+                    if (error.message.includes('Internal error opening backing store for indexedDB.open')) {
+                        degradePerformance(error);
+                    }
+                }
+
+                reject(error);
```

Comment:
It doesn't every look like anything is using this rejection. If so, maybe it should be removed?

Reply:
@tgolen I think no, it's better to have it here

The function `tryOrDegradePerformance` just acts as an interceptor - it simply executes the function and in case of a specific exception just substitues a provider. And when we develop such generic functions it's important to propagate exceptions further.

And this function keeps full backward compatibility - if `getItem` or any other function would throw exception before, then it would be thrown further (and potentially we may have a code that handles these errors). So I'd like to keep this line and propagate exceptions and don't silently swallow them.